javalib/src/main/scala/java/util/concurrent/CopyOnWriteArrayList.scala

```diff
@@ -325,19 +337,30 @@ class CopyOnWriteArrayList[E <: AnyRef] private (private var inner: js.Array[E])
     override protected def innerSet(index: Int, elem: E): Unit =
       self.innerSet(fromIndex + index, elem)
 
-    override protected def innerSplice(index: Int, deleteCount: Int,
-        items: E*): js.Array[E] = {
-      changeSize(items.size - deleteCount)
-      self.innerSplice(fromIndex + index, deleteCount, items: _*)
+    override protected def innerPush(elem: E): Unit = {
+      changeSize(1)
+      self.innerInsert(toIndex - 1, elem)
     }
 
-    override protected def innerPush(elem: E): Unit = {
-      if (toIndex < self.size) {
-        innerSplice(size, 0, elem)
-      } else {
-        changeSize(1)
-        self.innerPush(elem)
-      }
+    override protected def innerInsert(index: Int, elem: E): Unit = {
+      changeSize(1)
+      self.innerInsert(fromIndex + index, elem)
+    }
+
+    override protected def innerInsertMany(index: Int,
+        items: Collection[_ <: E]): Unit = {
+      changeSize(items.size())
+      self.innerInsertMany(fromIndex + index, items)
+    }
+
+    override protected def innerRemove(index: Int): E = {
+      changeSize(-1)
+      self.innerRemove(fromIndex + index)
+    }
+
+    override protected def innerRemoveMany(index: Int, count: Int): Unit = {
+      changeSize(-count)
+      self.innerRemoveMany(index, count)
```

Comment:
I don't necessarily disagree, but the strategy existed before this PR, and changing it would require a separate thought process which IMO does not belong in this PR.

It is correct not to call `copyIfNeeded()` because that is called by the public methods such as `add()` and `remove()` before they call the `inner` methods.

Reply:
Ah, right. Now I understand how this thing works... :-/ fair enough.